service/Core/Pipeline/InProcessPipelineOrchestrator.cs

```diff
@@ -165,25 +165,57 @@ public override async Task RunPipelineAsync(DataPipeline pipeline, CancellationT
 
             if (!this._handlers.TryGetValue(currentStepName, out var stepHandler))
             {
-                throw new OrchestrationException($"No handlers found for step '{currentStepName}'");
+                pipeline.LastUpdate = DateTimeOffset.UtcNow;
+                pipeline.Failed = true;
+                pipeline.Logs = $"No handler found for step '{currentStepName}'";
+                await this.UpdatePipelineStatusAsync(pipeline, cancellationToken).ConfigureAwait(false);
+
+                this.Log.LogError("No handler found for step '{0}'", currentStepName);
+                throw new OrchestrationException($"No handler found for step '{currentStepName}'");
             }
 
-            // Run handler
-            (bool success, DataPipeline updatedPipeline) = await stepHandler
-                .InvokeAsync(pipeline, this.CancellationTokenSource.Token)
-                .ConfigureAwait(false);
-            if (success)
+            try
             {
+                // Run handler
+                (bool success, DataPipeline updatedPipeline) = await stepHandler
+                    .InvokeAsync(pipeline, this.CancellationTokenSource.Token)
+                    .ConfigureAwait(false);
+
                 pipeline = updatedPipeline;
                 pipeline.LastUpdate = DateTimeOffset.UtcNow;
-                this.Log.LogInformation("Handler '{0}' processed pipeline '{1}/{2}' successfully", currentStepName, pipeline.Index, pipeline.DocumentId);
-                pipeline.MoveToNextStep();
-                await this.UpdatePipelineStatusAsync(pipeline, cancellationToken).ConfigureAwait(false);
+
+                if (success)
+                {
+                    this.Log.LogInformation("Handler '{0}' processed pipeline '{1}/{2}' successfully", currentStepName, pipeline.Index, pipeline.DocumentId);
+                    pipeline.MoveToNextStep();
+                    await this.UpdatePipelineStatusAsync(pipeline, cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    pipeline.Failed = true;
+                    pipeline.Logs = string.Join(", ", pipeline.Files.Where(f => f.LogEntries is not null).SelectMany(f => f.LogEntries!).Select(l => $"{l.Source}: {l.Text}"));
+                    await this.UpdatePipelineStatusAsync(pipeline, cancellationToken).ConfigureAwait(false);
+
+                    this.Log.LogError("Handler '{0}' failed to process pipeline '{1}/{2}'", currentStepName, pipeline.Index, pipeline.DocumentId);
+                    throw new OrchestrationException($"Pipeline error, step {currentStepName} failed");
+                }
             }
-            else
+            catch (Exception ex)
```

Comment:
ditto, where is the exception coming from? e.g. if `UpdatePipelineStatusAsync` is failing in the `try` block it will likely fail also in the `catch` block